src/adapter/src/coord/timestamp_selection.rs

```diff
@@ -161,14 +158,22 @@ impl<S: Append + 'static> Coordinator<S> {
             candidate.advance_by(since.borrow());
         }
 
-        if use_timestamp_oracle {
-            let timeline = timeline
-                .as_ref()
-                .expect("timeline is always present when using the timestamp oracle");
-            let timestamp_oracle = self.get_timestamp_oracle(timeline);
-            oracle_read_ts = Some(timestamp_oracle.read_ts());
-            candidate.join_assign(&oracle_read_ts.unwrap());
-        } else if when.advance_to_upper() {
+        if let Some(timeline) = &timeline {
+            if (isolation_level == &vars::IsolationLevel::StrictSerializable
+                && when.can_advance_to_global_ts())
+                || when.must_advance_to_global_ts()
+            {
+                let timestamp_oracle = self.get_timestamp_oracle(timeline);
+                oracle_read_ts = Some(timestamp_oracle.read_ts());
+                candidate.join_assign(&oracle_read_ts.unwrap());
+            } else if when.can_advance_to_upper() {
+                candidate.join_assign(&largest_not_in_advance_of_upper);
```

Comment:
This function is called like this 3 times, and the logic to follow why it should be called is hard to understand. Specifically, I'm trying to answer the question "when isn't this function called?" and can't do it convincingly. An idea that I don't like: make a `let advance_to_upper: bool = when.can_advance_to_upper()` that gets set to `false` in the oracle ts case, then do `if advance_to_upper|| when.must_advance_to_upper() { candidate.join_assign(&largest_not_in_advance_of_upper); }`. This would make it so advancing to an upper happens in exactly one place, and will maybe ease understanding of the design we're trying to achieve.

Also, I no longer recall why the if block here was "either do use_timestamp_oracle or use the upper", and wished there were a comment describing why we don't care about the upper in the oracle ts case.

Reply:
I tried to clean this up and add comments, let me know what you think of it now.